source/lac/scalapack.cc

```diff
@@ -0,0 +1,944 @@
+   * https://www.ibm.com/support/knowledgecenter/SSNR5K_4.2.0/com.ibm.cluster.pessl.v4r2.pssl100.doc/am6gr_lsyev.htm#lsyev
+   */
+  void pdsyev_(const char *jobz, const char *uplo, const int *m, double *A, const int *ia, const int *ja, int *desca, double *w,
+               double *z, const int *iz, const int *jz, int *descz, double *work, const int *lwork, int *info);
+
+  /*
+   * pdlacpy copies all or a part of a distributed matrix A to another
+   * distributed matrix B. No communication is performed, pdlacpy
+   * performs a local copy sub(A) := sub(B), where sub(A) denotes
+   * A(ia:ia+m-1,ja:ja+n-1) and sub(B) denotes B(ib:ib+m-1,jb:jb+n-1)
+   *
+   */
+  void pdlacpy_(const char *uplo, const int *m, const int *n, double *A, const int *ia, const int *ja, int *desca,
+                double *B, const int *ib, const int *jb, int *descb);
+}
+
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace
+{
+  /**
+   * Internal function to determine dimension of process grid based on the total
+   * number of cores, matrix dimensions and matrix block sizes
+   *
+   * Amesos heuristics:
+   * https://github.com/trilinos/Trilinos/blob/master/packages/amesos/src/Amesos_Scalapack.cpp#L142-L166
+   *
+   * Elemental default grid: El::Grid::Grid(mpi_comm,...)
+   * https://github.com/elemental/Elemental/blob/master/src/core/Grid.cpp#L67-L91
+   */
+  inline
+  std::pair<int,int> choose_the_processor_grid(MPI_Comm mpi_comm, const unsigned int m, const unsigned int n,
+                                               const unsigned int block_size_m, const unsigned int block_size_n)
+  {
+    // Few notes from the ScaLAPACK user guide:
+    // It is possible to predict the best grid shape given the number of processes available:
+    // Pr x Pc <= P
+    // This, however, depends on the task to be done.
+    // LU , QR and QL factorizations perform better for “flat” process grids (Pr < Pc )
+    // For large N, Pc = 2*Pr is a good choice, whereas for small N, one should choose small Pr
+    // Square or near square grids are more optimal for Cholesky factorization.
+    // LQ and RQ factorizations take advantage of “tall” grids (Pr > Pc )
+
+    // Below we always try to create 2D processor grids:
+
+    int n_processes;
+    MPI_Comm_size(mpi_comm, &n_processes);
+
+    // Get the total number of cores we can occupy in a rectangular dense matrix
+    // with rectangular blocks when every core owns only a single block:
+    const int n_processes_heuristic = int(std::ceil((1.*m)/block_size_m))*
+                                      int(std::ceil((1.*n)/block_size_n));
+    const int Np = std::min(n_processes_heuristic, n_processes);
+
+    // Now we need to split Np into  Pr x Pc. Assume we know the shape/ratio
+    // Pc =: ratio * Pr
+    // therefore
+    // Np = Pc * Pc / ratio
+    // for quadratic matrices the ratio equals 1
+    const double ratio = double(n)/m;
+    int Pc = std::sqrt(ratio * Np);
```

Comment:
is this conversion using `round` or `floor`? Maybe make it explicit?

Reply:
i would say `floor`, i will update.